### PR TITLE
[R4R] Implement bep 131

### PR DESCRIPTION
**Pull request by dean65**
Ref: https://github.com/bnb-chain/BEPs/pull/131

### CODE REVIEW

File: contracts/BSCValidatorSet.sol [hunk 1]
```diff
@@ -67,6 +67,11 @@ contract BSCValidatorSet is IBSCValidatorSet, System, IParamSubscriber, IApplica
   // key is the `consensusAddress` of `Validator`,
   mapping(address => MaintainInfo) public maintainInfoMap;
 
+  // BEP-131 candidate validator
+  uint256 public numOfCabinets = 21;
```
Comment (guagualvcha) [inline contracts/BSCValidatorSet.sol:71]: When we do upgrade, we can only upgrade the code of smart contract, so the storage of this contract will not change, it means `numOfCabinets` will be zero after we upgrade it. So you have to assume `numOfCabinets`, `maxNumOfCandidates` and `maxNumOfWorkingCandidates` is zero before we change the value of them through governance .

File: contracts/BSCValidatorSet.sol [hunk 2]
```diff
@@ -344,6 +349,36 @@ contract BSCValidatorSet is IBSCValidatorSet, System, IParamSubscriber, IApplica
     return CODE_OK;
   }
 
+  function swap(address[] memory set, uint i, uint j) internal pure {
+    address tmp = set[i];
+    set[i]=set[j];
+    set[j]=tmp;
+  }
+
+  function shuffle(address[] memory validators, uint256 epochNumber, uint start, uint limit, uint modNumber) internal pure {
```
Comment (guagualvcha) [inline contracts/BSCValidatorSet.sol:358]: We can delete `modNumber` here, it should be length of `validators`

File: contracts/BSCValidatorSet.sol [hunk 3]
```diff
@@ -344,6 +349,36 @@ contract BSCValidatorSet is IBSCValidatorSet, System, IParamSubscriber, IApplica
...
+    }
+  }
+
+  function getMiningValidators(uint256 epochNumber) public view returns(address[] memory) {
+    address[] memory validators = getValidators();
+    if (validators.length <= numOfCabinets) {
+      return validators;
+    }
+    
+    shuffle(validators, epochNumber, 0, maxNumOfWorkingCandidates, numOfCabinets);
```
Comment (guagualvcha) [inline contracts/BSCValidatorSet.sol:373]: index out of range

Comment (guagualvcha) [inline contracts/BSCValidatorSet.sol:373]: Should do reverse shuffle

File: contracts/BSCValidatorSet.sol [hunk 4]
```diff
@@ -344,6 +349,36 @@ contract BSCValidatorSet is IBSCValidatorSet, System, IParamSubscriber, IApplica
...
+    shuffle(validators, epochNumber, numOfCabinets-maxNumOfWorkingCandidates, maxNumOfWorkingCandidates, validators.length-numOfCabinets+maxNumOfWorkingCandidates);
+    address[] memory miningValidators = new address[](numOfCabinets);
+    for (uint i=0;i<numOfCabinets;i++) {
+      miningValidators[i] = validators[i];
+    }
+    return miningValidators;
```
Comment (guagualvcha) [inline contracts/BSCValidatorSet.sol:379]: Try if slice works on solidity

File: contracts/BSCValidatorSet.sol [hunk 5]
```diff
@@ -344,6 +349,36 @@ contract BSCValidatorSet is IBSCValidatorSet, System, IParamSubscriber, IApplica
     return CODE_OK;
   }
 
+  function swap(address[] memory set, uint i, uint j) internal pure {
```
[the remark this answers is not in the document]
Comment (dean65) [reply contracts/BSCValidatorSet.sol]: yes, I have tested, and update `set` => `addresses`

File: contracts/BSCValidatorSet.sol (hunk 3 above)
Comment (dean65) [reply contracts/BSCValidatorSet.sol]: fixed

Comment (dean65) [reply contracts/BSCValidatorSet.sol]: fixed

File: contracts/BSCValidatorSet.sol (hunk 4 above)
Comment (dean65) [reply contracts/BSCValidatorSet.sol:392]: It seems that getting slices in solidity does not support the `memory` type, I try to return validator[0:n] will get compiling error

File: contracts/BSCValidatorSet.sol [hunk 6]
```diff
@@ -344,6 +352,46 @@ contract BSCValidatorSet is IBSCValidatorSet, System, IParamSubscriber, IApplica
...
 
+  function swap(address[] memory addresses, uint i, uint j) internal pure {
+    address tmp = addresses[i];
+    addresses[i]=addresses[j];
+    addresses[j]=tmp;
+  }
+
+  function shuffle(address[] memory validators, uint256 epochNumber, uint startIdx, uint offset, uint limit, uint modNumber) internal pure {
+    for (uint i = 0; i<limit; i++) {
+      uint random = uint(keccak256(abi.encodePacked(epochNumber, validators[startIdx+i]))) % modNumber;
```
Comment (gothery001) [inline contracts/BSCValidatorSet.sol:363]: It is best not to give `Validators` the opportunity to change the random factor

**Review by gothery001 — CHANGES_REQUESTED** [review]

File: contracts/BSCValidatorSet.sol (hunk 6 above)
Comment (dean65) [reply contracts/BSCValidatorSet.sol]: updated

**Review by forcodedancing — DISMISSED** [review]

File: contracts/BSCValidatorSet.sol [hunk 7]
```diff
@@ -67,6 +67,14 @@ contract BSCValidatorSet is IBSCValidatorSet, System, IParamSubscriber, IApplica
   // key is the `consensusAddress` of `Validator`,
   mapping(address => MaintainInfo) public maintainInfoMap;
 
+  // BEP-131 candidate validator
+  uint256 public constant INIT_NUM_OF_CABINETS = 21;
```
Comment (realuncle) [inline contracts/BSCValidatorSet.sol:71]: Please move the constant to the top of the file

File: contracts/BSCValidatorSet.sol [hunk 8]
```diff
@@ -344,6 +352,46 @@ contract BSCValidatorSet is IBSCValidatorSet, System, IParamSubscriber, IApplica
...
+    uint256 _maxNumOfWorkingCandidates = maxNumOfWorkingCandidates;
+    uint256 _numOfCabinets = numOfCabinets;
+    if (_maxNumOfWorkingCandidates == 0 ){
+      _maxNumOfWorkingCandidates = INIT_MAX_NUM_OF_WORKING_CANDIDATES;
+    }
+    if (_numOfCabinets == 0 ){
+      _numOfCabinets = INIT_NUM_OF_CABINETS;
+    }
+
+    address[] memory validators = getValidators();
```
Comment (realuncle) [inline contracts/BSCValidatorSet.sol:376]: Please make sure the return of `getValidators` is sorted by voting power.

File: contracts/BSCValidatorSet.sol [hunk 9]
```diff
@@ -344,6 +352,46 @@ contract BSCValidatorSet is IBSCValidatorSet, System, IParamSubscriber, IApplica
...
+      uint random = uint(keccak256(abi.encodePacked(epochNumber, startIdx+i))) % modNumber;
+      if ( (startIdx+i) != (offset+random) ) {
+        address tmp = validators[startIdx+i];
+        validators[startIdx+i] = validators[offset+random];
+        validators[offset+random] = tmp;
+      }
+    }
+  }
+
+  function getMiningValidators(uint256 epochNumber) public view returns(address[] memory) {
```
Comment (realuncle) [inline contracts/BSCValidatorSet.sol:366]: I think we can do not pass `epochNumber` as param, but calculate by `currentBlockHeight % EPOCH`, `EPOCH=200`.

Comment (realuncle) [reply contracts/BSCValidatorSet.sol:366]: We wont change `EPOCH ` in near future, even we change it, it is a big hardfork.

File: contracts/BSCValidatorSet.sol [hunk 10]
```diff
@@ -478,6 +526,26 @@ contract BSCValidatorSet is IBSCValidatorSet, System, IParamSubscriber, IApplica
...
+      if (_maxNumOfCandidates == 0) {
+        _maxNumOfCandidates = INIT_MAX_NUM_OF_CANDIDATES;
+      }
+      require(newMaxNumOfWorkingCandidates <= _maxNumOfCandidates, "the maxNumOfCandidates must be not greater than maxNumOfCandidates");
+      maxNumOfWorkingCandidates = newMaxNumOfWorkingCandidates;
+    } else if (Memory.compareStrings(key, "maxNumOfCandidates")) {
+      require(value.length == 32, "length of maxNumOfCandidates mismatch");
+      uint256 newMaxNumOfCandidates = BytesToTypes.bytesToUint256(32, value);
+      require(newMaxNumOfCandidates > 0, "the maxNumOfCandidates must be not less than 0");
+      maxNumOfCandidates = newMaxNumOfCandidates;
```
Comment (realuncle) [inline contracts/BSCValidatorSet.sol:543]: If we follow the same code stye, you should init maxNumOfCandidates it the value is zero. But I suggest delete the init of  _maxNumOfCandidates too.

File: test/test-maintenance/BSCValidatorSet.spec.ts [hunk 11]
```diff
@@ -564,4 +564,53 @@ describe('BSCValidatorSet', () => {
...
+    if ((validators.length - numOfCabinets) < maxNumOfWorkingCandidates){
+      maxNumOfWorkingCandidates = validators.length - numOfCabinets;
+    } 
+    for (var i=epochNumber;i <5;i++){
+      var miningValidators = await validatorSet.getMiningValidators(i);
+      var exceptValues = miningValidators.slice(0,numOfCabinets-maxNumOfWorkingCandidates);
+      for (var j=0;j<numOfCabinets-maxNumOfWorkingCandidates;j++){
+        expect(exceptValues).to.includes(miningValidators[j]);
+      }
+      expect(miningValidators).to.deep.eq(await validatorSet.getMiningValidators(i));
```
Comment (realuncle) [inline test/test-maintenance/BSCValidatorSet.spec.ts:613]: Is not it definitely to be true?

File: test/test-maintenance/BSCValidatorSet.spec.ts [hunk 12]
```diff
@@ -564,4 +564,53 @@ describe('BSCValidatorSet', () => {
...
+    var epochNumber = 0;
+    var numOfCabinets = 17;
+
+    if ((validators.length - numOfCabinets) < maxNumOfWorkingCandidates){
+      maxNumOfWorkingCandidates = validators.length - numOfCabinets;
+    } 
+    for (var i=epochNumber;i <5;i++){
+      var miningValidators = await validatorSet.getMiningValidators(i);
+      var exceptValues = miningValidators.slice(0,numOfCabinets-maxNumOfWorkingCandidates);
+      for (var j=0;j<numOfCabinets-maxNumOfWorkingCandidates;j++){
```
Comment (realuncle) [inline test/test-maintenance/BSCValidatorSet.spec.ts:610]: Is not it definitely to be true?

We need to expect that  miningValidators to include `numOfCabinets-maxNumOfWorkingCandidates` canbinates

File: contracts/BSCValidatorSet.sol (hunk 8 above)
Comment (realuncle) [reply contracts/BSCValidatorSet.sol:381]: I have double checked it. It is not strickly in order...     Please search `It is ok that the validatorSet is not in order.` in the repo and check the function `doUpdateState`

Comment (dean65) [reply contracts/BSCValidatorSet.sol:381]: I changed the logic of the felony function to ensure the order of the currentValidatorSet , please check

File: contracts/BSCValidatorSet.sol (hunk 9 above)
Comment (dean65) [reply contracts/BSCValidatorSet.sol]: updated

File: contracts/BSCValidatorSet.sol (hunk 10 above)
Comment (dean65) [reply contracts/BSCValidatorSet.sol:556]: updated

File: test/test-maintenance/BSCValidatorSet.spec.ts (hunk 11 above)
Comment (dean65) [reply test/test-maintenance/BSCValidatorSet.spec.ts]: updated

File: test/test-maintenance/BSCValidatorSet.spec.ts (hunk 12 above)
Comment (dean65) [reply test/test-maintenance/BSCValidatorSet.spec.ts]: updated

File: contracts/BSCValidatorSet.sol (hunk 10 above)
Comment (dean65) [reply contracts/BSCValidatorSet.sol:556]: In the initial state,`maxNumOfCandidates ` and `maxNumOfWorkingCandidates ` will be 0, which means that candidate validators will not join the consensus

File: contracts/BSCValidatorSet.sol [hunk 13]
```diff
@@ -312,7 +319,46 @@ contract BSCValidatorSet is IBSCValidatorSet, System, IParamSubscriber, IApplica
...
+    address[] memory validators = getValidators();
+    if (validators.length <= _numOfCabinets) {
+      return validators;
+    }
+
+    if ((validators.length - _numOfCabinets) < _maxNumOfWorkingCandidates){
+      _maxNumOfWorkingCandidates = validators.length - _numOfCabinets;
+    }
+    if (_maxNumOfWorkingCandidates > 0) {
+      uint256 epochNumber = block.number & EPOCH;
```
Comment (realuncle) [inline contracts/BSCValidatorSet.sol]: I suppose to be `      uint256 epochNumber = block.number / EPOCH;`

File: test/BSCValidatorSet.js [hunk 14]
```diff
@@ -600,6 +600,133 @@ contract('BSCValidatorSet', (accounts) => {
...
+    except = await validatorSetInstance.maxNumOfWorkingCandidates.call();
+    assert.equal(web3.utils.toBN(except).eq(web3.utils.toBN(2)), true, "wrong maxNumOfWorkingCandidates");
+
+    if ((validators.length - numOfCabinets) < maxNumOfWorkingCandidates){
+      maxNumOfWorkingCandidates = validators.length - numOfCabinets;
+    } 
+    
+    miningValidators = await validatorSetInstance.getMiningValidators.call();
+    let exceptValues = validators.slice(0,numOfCabinets);
+    for (var j=0;j<numOfCabinets-maxNumOfWorkingCandidates;j++){
```
Comment (realuncle) [inline test/BSCValidatorSet.js:726]: What about the `miningValidators[numOfCabinets-maxNumOfWorkingCandidates: numOfCabinets]`, can you come up with any idea to ensure that candidates can be one of them?

File: contracts/BSCValidatorSet.sol [hunk 15]
```diff
@@ -413,6 +459,25 @@ contract BSCValidatorSet is IBSCValidatorSet, System, IParamSubscriber, IApplica
       require(newBurnRatio <= BURN_RATIO_SCALE, "the burnRatio must be no greater than 10000");
       burnRatio = newBurnRatio;
       burnRatioInitialized = true;
+    } else if (Memory.compareStrings(key, "maxNumOfWorkingCandidates")) {
+      require(value.length == 32, "length of maxNumOfWorkingCandidates mismatch");
+      uint256 newMaxNumOfWorkingCandidates = BytesToTypes.bytesToUint256(32, value);
+      require(newMaxNumOfWorkingCandidates >= 0, "the maxNumOfWorkingCandidates must be not less than 0");
```
Comment (NashBC) [inline contracts/BSCValidatorSet.sol]: No need, always true.

File: contracts/BSCValidatorSet.sol [hunk 16]
```diff
@@ -413,6 +459,25 @@ contract BSCValidatorSet is IBSCValidatorSet, System, IParamSubscriber, IApplica
...
+      require(newMaxNumOfWorkingCandidates <= maxNumOfCandidates, "the maxNumOfWorkingCandidates must be not greater than maxNumOfCandidates");
+      maxNumOfWorkingCandidates = newMaxNumOfWorkingCandidates;
+    } else if (Memory.compareStrings(key, "maxNumOfCandidates")) {
+      require(value.length == 32, "length of maxNumOfCandidates mismatch");
+      uint256 newMaxNumOfCandidates = BytesToTypes.bytesToUint256(32, value);
+      require(newMaxNumOfCandidates >= 0, "the maxNumOfCandidates must be not less than 0");
```
Comment (NashBC) [inline contracts/BSCValidatorSet.sol]: No need.

File: contracts/BSCValidatorSet.sol [hunk 17]
```diff
@@ -312,7 +319,46 @@ contract BSCValidatorSet is IBSCValidatorSet, System, IParamSubscriber, IApplica
...
+  }
+
+  function getMiningValidators() public view returns(address[] memory) {
+    uint256 _maxNumOfWorkingCandidates = maxNumOfWorkingCandidates;
+    uint256 _numOfCabinets = numOfCabinets;
+    if (_numOfCabinets == 0 ){
+      _numOfCabinets = INIT_NUM_OF_CABINETS;
+    }
+
+    address[] memory validators = getValidators();
```
Comment (NashBC) [inline contracts/BSCValidatorSet.sol:381]: Shall we ensure the validatorSet is sorted in checkValidatorSet()? If the validatorSet is passed not sorted, we should reject it in the first place.

File: contracts/BSCValidatorSet.sol [hunk 18]
```diff
@@ -413,6 +459,25 @@ contract BSCValidatorSet is IBSCValidatorSet, System, IParamSubscriber, IApplica
...
+      uint256 newMaxNumOfCandidates = BytesToTypes.bytesToUint256(32, value);
+      require(newMaxNumOfCandidates >= 0, "the maxNumOfCandidates must be not less than 0");
+      maxNumOfCandidates = newMaxNumOfCandidates;
+      if (maxNumOfWorkingCandidates > maxNumOfCandidates) {
+        maxNumOfWorkingCandidates = maxNumOfCandidates;
+      }
+    } else if (Memory.compareStrings(key, "numOfCabinets")) {
+      require(value.length == 32, "length of numOfCabinets mismatch");
+      uint256 newNumOfCabinets = BytesToTypes.bytesToUint256(32, value);
+      require(newNumOfCabinets > 0, "the numOfCabinets must be greater than 0");
```
Comment (NashBC) [inline contracts/BSCValidatorSet.sol:563]: require(newNumOfCabinets <= MAX_NUM_OF_VALIDATORS, ...)

File: contracts/BSCValidatorSet.sol (hunk 17 above)
Comment (realuncle) [reply contracts/BSCValidatorSet.sol:381]: It is guaranteed by `BNB beacon chain`.

**Review by realuncle — APPROVED** [review]

**Review by NashBC — APPROVED** [review]

**Review by gothery001 — APPROVED** [review]

**Review by axljoker03 — APPROVED** [review]

**Review by mehdi2021mp — APPROVED** [review]

**Review by mehdi2021mp — CHANGES_REQUESTED** [review]
In general, I want  I have to sign an address or a contract and I have to do this signing from your wallet. Thank you

**Comment by mehdi2021mp** [comment]
In general, I want  I have to sign an address or a contract and I have to do this signing from your wallet. Thank you

**Pull request merged**